### PR TITLE
Add Scheduler validation check for redeclared policy configs

### LINKED ISSUE #83472: Duplicate Custom Predicates/Priorities
Currently custom predicate and priorities can be created through the [Policy API](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/v1/types.go#L51-L71). 

The problem is that a custom predicate/priority, like ServiceAffinity, can be [listed](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/v1/types.go#L32-L35) multiple times in the Policy with different names, possibly with different arguments. 

This semantic (i.e., specifying multiple instances of the same custom predicate with different arguments) is only useful if different ```Weight```s were specified because in all cases of custom predicates/priorities the arguments are merge-able (for example, ```Labels``` from multiple CheckServiceAffinity custom predicates can be merged into one set).

In the world of framework plugins, we don't currently have a way to maintain those semantics. [```PluginConfig```](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/types.go#L98), which specifies plugins' arguments is separate from list that specifies which [```Plugins```](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/apis/config/types.go#L91) to run.

The question here: do we really want to maintain the old semantics of custom predicates and priorities? in other words, do we need to change the ComponentConfig API to somehow allow specifying multiple instances of the same Plugin with different arguments/weights, or should we consider that this is not supported?

I always lean towards simplifying things, so my suggestion is to not support that, but I would like to know if there are actual deployments that rely on this.

/sig scheduling

/cc @bsalamat @misterikkit @Huang-Wei @k82cn @ravisantoshgudimetla

### PR DESCRIPTION
**Pull request by damemi**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
The ability to re-declare custom policy configs was not entirely intentional, and this adds validation to disallow that in the new framework

**Which issue(s) this PR fixes**:
Fixes #83472

```release-note
Scheduler policy configs can no longer be declared multiple times
```

### CODE REVIEW

**Comment by BenTheElder** [comment]
/retest
https://github.com/kubernetes-sigs/kind/pull/951 for `pull-kubernetes-e2e-kind` issue (worked around)

**Comment by damemi** [comment]
> Hi @damemi
> From the issue, it seems that we need a validator for the old path, rather than the framework path.
> https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/validation/validation.go#L34

@alculquicondor Agree that I did take the wrong direction with this PR, but I spoke with @ahg-g and it seems like we may want to do this check when actually registering the custom policies, as in here: https://github.com/kubernetes/kubernetes/blob/6b2b5f25fd5f12b251f9a7f8953da89c47741a95/pkg/scheduler/algorithm_factory.go#L368

What do you think? Can probably accomplish the same thing in either spot

**Comment by alculquicondor** [comment]
that's right, probably both should do the work. However, the validation.go file has more user-facing semantics.
WDYT @ahg-g?

**Comment by ahg-g** [comment]
> that's right, probably both should do the work. However, the validation.go file has more user-facing semantics.
> WDYT @ahg-g?

Aldo's suggestion SGTM.

**Comment by damemi** [comment]
Ok, updated this PR to do validation based on the arguments given in the policies in `validation.go`. Because we have a sort of polymorphism with the argument types, I couldn't think of a much cleaner way to check them, but please review and if you have any ideas for how to refactor this it would be helpful

File: pkg/scheduler/api/validation/validation.go [hunk 1]
```diff
@@ -66,6 +74,44 @@ func ValidatePolicy(policy schedulerapi.Policy) error {
...
+	}
+	priorities[priorityType] = priority
+	return nil
+}
+
+// validatePredicateRedeclared checks if any custom predicates have been declared multiple times in the policy config
+// by examining the specified predicate arguments
+func validatePredicateRedeclared(predicates map[string]schedulerapi.PredicatePolicy, predicate schedulerapi.PredicatePolicy) error {
+	var predicateType string
+	if predicate.Argument.LabelsPresence != nil {
```
Comment (liu-cong) [inline pkg/scheduler/api/validation/validation.go]: I think you need to check whether predicate.Argument is nil or not.

Currently `nil` argument is valid if the name of the predicate matches a predefined one (I don't know if this is desired or not), see
https://github.com/kubernetes/kubernetes/blob/e1f86e346062e36aa03fc53ea4fb05d4a61ef458/pkg/scheduler/algorithm_factory.go#L281

**Review by liu-cong — COMMENTED** [review]
Can you add a unit test for this

**Review by alculquicondor — COMMENTED** [review]
Unit tests, please :)

**Review by ahg-g — COMMENTED** [review]
Thanks Mike, can you please add a compatibility test as well.

File: pkg/scheduler/api/validation/validation_test.go [hunk 2]
```diff
@@ -102,6 +102,26 @@ func TestValidatePolicy(t *testing.T) {
...
+			policy: api.Policy{
+				Predicates: []api.PredicatePolicy{
+					{Name: "customPredicate1", Argument: &api.PredicateArgument{ServiceAffinity: &api.ServiceAffinity{Labels: []string{"label1"}}}},
+					{Name: "customPredicate2", Argument: &api.PredicateArgument{ServiceAffinity: &api.ServiceAffinity{Labels: []string{"label2"}}}},
+				},
+			},
+			expected: errors.New("Predicate 'customPredicate2' redeclares custom predicate 'ServiceAffinity', from:'customPredicate1'"),
+		},
+		{
+			name: "invalid redeclared custom priority",
```
Comment (alculquicondor) [inline pkg/scheduler/api/validation/validation_test.go]: nit: a different name from the previous one

Comment (damemi) [reply pkg/scheduler/api/validation/validation_test.go]: `invalid redeclared custom predicate` and `invalid redeclared custom priority`? They are testing the same thing, but for predicate/priority

**Pull request merged**